Locally invoked functions clear webpack compiled code needed for functions which are called from locally invoked function

<!--
1. Please check if an issue already exists so there are no duplicates
2. Fill out the whole template so we have a good overview on the issue
3. Do not remove any section of the template. If something is not applicable leave it empty but leave it in the Issue
4. Please follow the template, otherwise we'll have to ask you to update it
-->

# This is a Bug Report

## Description
When invoking a local function which publishes to an SNS topic, the called lambda function which is subscribed to the SNS topic fails due to the webpack compiled code being removed during the local invocation. This issue is only during local development, and using serverless-offline-sns.

Example repository to highlight the issue: https://github.com/todda00/serverless-offline-sns-example/tree/webpack-local-invoke-issue

Steps to reproduce:

Simply run
```bash
npm install
npm start
```

And then go to http://localhost:3000/ping to trigger the event, you should see "pong" appear in the console.

To see the issue with webpack, open a new command line (while serverless offline is still running) and run:

```bash
serverless invoke local -f ping
```

You will see the trace error in the terminal where serverless offline is running.

The .webpack folder is cleared out during validation and is only compiling the locally invoked function (ping), but the running offline instance needs the pong handler, which has been removed.

For bug reports:
* What went wrong?
compiled modules of other functions are removed during local invocations
* What did you expect should have happened?
both functions to be able to be ran in sequence without webpack removing the compiled code
* What was the config you used?
see example repository: https://github.com/todda00/serverless-offline-sns-example/tree/webpack-local-invoke-issue
* What stacktrace or error message from your provider did you see?
```
Error: Cannot find module '/Users/toddhumphrey/Repos/serverless-offline-sns-example/.webpack/service/pong'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at ServerlessOfflineSns.createHandler (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/serverless-offline-sns/dist/index.js:199:23)
    at /Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/serverless-offline-sns/dist/index.js:178:107
    at /Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/serverless-offline-sns/dist/sns-adapter.js:141:29
    at Layer.handle [as handle_request] (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/layer.js:95:5)
    at next (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/layer.js:95:5)
    at /Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/index.js:281:22
    at Function.process_params (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/index.js:335:12)
    at next (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/index.js:275:10)
    at /Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/serverless-offline-sns/dist/sns-server.js:44:13
    at Layer.handle [as handle_request] (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/index.js:317:13)
    at /Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/index.js:335:12)
    at next (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/index.js:275:10)
    at urlencodedParser (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/body-parser/lib/types/urlencoded.js:82:7)
    at Layer.handle [as handle_request] (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/index.js:317:13)
    at /Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/index.js:335:12)
    at next (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/express/lib/router/index.js:275:10)
    at /Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/body-parser/lib/read.js:130:5
    at invokeCallback (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/raw-body/index.js:224:16)
    at done (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/raw-body/index.js:213:7)
    at IncomingMessage.onEnd (/Users/toddhumphrey/Repos/serverless-offline-sns-example/node_modules/raw-body/index.js:273:7)
    at emitNone (events.js:86:13)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:974:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickDomainCallback (internal/process/next_tick.js:122:9)
```

Similar or dependent issue(s):

## Additional Data

* ***Serverless-Webpack Version you're using***: 4.0.0
* ***Webpack version you're using***: 3.8.1
* ***Serverless Framework Version you're using***: 1.24.1
* ***Operating System***: MacOS Sierra v 10.12.6
* ***Stack Trace (if available)***:
see trace above

--- a/README.md
+++ b/README.md
@@ -422,6 +422,19 @@ All options that are supported by invoke local can be used as usual:
 
 > :exclamation: The old `webpack invoke` command has been disabled.
 
+#### Run a function with an existing compiled output (--no-build)
+
+On CI systems it is likely that you'll run multiple integration tests with `invoke local`
+sequentially. To improve this, you can do one compile and run multiple invokes on the
+compiled output -  it is not necessary to compile again before each and every invoke.
+
+```bash
+$ serverless webpack
+$ serverless invoke local --function <function-name-1> --no-build
+$ serverless invoke local --function <function-name-2> --no-build
+...
+```
+
 ### Run a function locally on source changes
 
 Or to run a function every time the source files change use the `--watch` option

--- a/index.js
+++ b/index.js
@@ -112,9 +112,14 @@ class ServerlessWebpack {
       'before:invoke:local:invoke': () => BbPromise.bind(this)
         .then(() => {
           lib.webpack.isLocal = true;
+          // --no-build override
+          if (this.options.build === false) {
+            this.skipCompile = true;
+          }
+
           return this.serverless.pluginManager.spawn('webpack:validate');
         })
-        .then(() => this.serverless.pluginManager.spawn('webpack:compile'))
+        .then(() => this.skipCompile ? BbPromise.resolve() : this.serverless.pluginManager.spawn('webpack:compile'))
         .then(this.prepareLocalInvoke),
 
       'after:invoke:local:invoke': () => BbPromise.bind(this)
@@ -181,7 +186,7 @@ class ServerlessWebpack {
           lib.webpack.isLocal = true;
         })
         .then(this.prepareStepOfflineInvoke)
-        .then(this.compile)
+        .then(() => this.serverless.pluginManager.spawn('webpack:compile')),
     };
   }
 }

--- a/index.test.js
+++ b/index.test.js
@@ -357,7 +357,8 @@ describe('ServerlessWebpack', () => {
             .then(() => {
               expect(ServerlessWebpack.lib.webpack.isLocal).to.be.true;
               expect(slsw.prepareStepOfflineInvoke).to.have.been.calledOnce;
-              expect(slsw.compile).to.have.been.calledOnce;
+              expect(slsw.serverless.pluginManager.spawn).to.have.been.calledOnce;
+              expect(slsw.serverless.pluginManager.spawn).to.have.been.calledWithExactly('webpack:compile');
               return null;
             });
           });

--- a/index.test.js
+++ b/index.test.js
@@ -200,6 +200,17 @@ describe('ServerlessWebpack', () => {
               return null;
             });
           });
+
+          it('should skip compile if requested', () => {
+            slsw.options.build = false;
+            return expect(slsw.hooks['before:invoke:local:invoke']()).to.be.fulfilled
+            .then(() => {
+              expect(slsw.serverless.pluginManager.spawn).to.have.been.calledOnce;
+              expect(slsw.serverless.pluginManager.spawn).to.have.been.calledWithExactly('webpack:validate');
+              expect(slsw.prepareLocalInvoke).to.have.been.calledOnce;
+              return null;
+            });
+          });
         }
       },
       {

--- a/lib/validate.js
+++ b/lib/validate.js
@@ -143,6 +143,15 @@ module.exports = {
       this.webpackConfig.output.path = path.join(this.serverless.config.servicePath, this.options.out);
     }
 
+    if (this.skipCompile) {
+      this.serverless.cli.log('Skipping build and using existing compiled output');
+      if (!fse.pathExistsSync(this.webpackConfig.output.path)) {
+        return BbPromise.reject(new this.serverless.classes
+          .Error('No compiled output found'));
+      }
+      this.keepOutputDirectory = true;
+    }
+
     if (!this.keepOutputDirectory) {
       this.options.verbose && this.serverless.cli.log(`Removing ${this.webpackConfig.output.path}`);
       fse.removeSync(this.webpackConfig.output.path);

--- a/tests/mocks/fs-extra.mock.js
+++ b/tests/mocks/fs-extra.mock.js
@@ -8,6 +8,7 @@ module.exports.create = sandbox => {
   const fsExtraMock = {
     copy: sandbox.stub().yields(),
     pathExists: sandbox.stub().yields(),
+    pathExistsSync: sandbox.stub().returns(false),
     removeSync: sandbox.stub()
   };
 

--- a/tests/validate.test.js
+++ b/tests/validate.test.js
@@ -53,6 +53,7 @@ describe('validate', () => {
 
   afterEach(() => {
     fsExtraMock.removeSync.reset();
+    fsExtraMock.pathExistsSync.reset();
     sandbox.restore();
   });
 
@@ -733,6 +734,38 @@ describe('validate', () => {
         });
       });
     });
+  });
 
+  describe('with skipped builds', () => {
+    it('should keep output directory', () => {
+      const testConfig = {
+        entry: 'test',
+        output: {},
+      };
+      const testServicePath = 'testpath';
+      module.serverless.config.servicePath = testServicePath;
+      _.set(module.serverless.service, 'custom.webpack.config', testConfig);
+      module.skipCompile = true;
+      fsExtraMock.pathExistsSync.returns(true);
+      return module
+      .validate()
+      .then(() => {
+        expect(module.keepOutputDirectory).to.be.true;
+        return null;
+      });
+    });
+
+    it('should fail without exiting output', () => {
+      const testConfig = {
+        entry: 'test',
+        output: {},
+      };
+      const testServicePath = 'testpath';
+      module.serverless.config.servicePath = testServicePath;
+      _.set(module.serverless.service, 'custom.webpack.config', testConfig);
+      module.skipCompile = true;
+      fsExtraMock.pathExistsSync.returns(false);
+      return expect(module.validate()).to.be.rejectedWith(/No compiled output found/);
+    });
   });
 });

